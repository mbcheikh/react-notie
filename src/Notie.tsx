import * as React from 'react';
import { css } from 'emotion';

export enum NotieLevel {
    INFO = 'INFO',
    ERROR = 'ERROR',
    SUCCESS = 'SUCCESS',
    WARN = 'WARN',
    FORCE = 'FORCE',
    CONFIRM = 'CONFIRM'
}

const colors: Record<NotieLevel, string> = {
    [NotieLevel.SUCCESS]: '#13ce66',
    [NotieLevel.ERROR]: '#ff4949',
    [NotieLevel.INFO]: '#20a0ff',
    [NotieLevel.WARN]: '#f5be4b',
    [NotieLevel.CONFIRM]: '#20a0ff',
    [NotieLevel.FORCE]: '#20a0ff'
};

//#region styled_components
export const notieContainer = css({
    position: 'fixed',
    width: '100%',
    textAlign: 'center',
    fontSize: '14px',
    color: '#fff',
    margin: '0 auto',
    whiteSpace: 'nowrap',
    border: 'none',
    textOverflow: 'ellipsis',
    left: 0,
    top: 0,
    padding: 0,
    pointerEvents: 'none',
    '&[open]': {
        pointerEvents: 'auto'
    }
});

export const notieMessage = css({
    padding: '20px'
});

export const notieChoices = css({
    display: 'flex',
    width: '100%'
});

export const notieButton = css({
    padding: '20px',
    flexGrow: 1,
    border: 'none',
    textAlign: 'center',
    cursor: 'pointer',
    outline: 'none'
});
//#endregion

/**
 * Settings for Notie
 */
export interface NotieSettings {
    /**
     * The message to be shown
     */
    message: string;
    /**
     * The type of alert-box to shown
     */
    level: NotieLevel;
    /**
     * Text to be shown for 'OK' button
     *
     * @default 'OK'
     */
    okBtnText?: string;
    /**
     * Text to be shown for 'Cancel' button
     *
     * @default 'Cancel'
     */
    cancelBtnText?: string;
    /**
     * The time (in ms) after which the alert-box disappears
     * @default 5000
     */
    ttl?: number;
}

export type NotieCallback = (...args: any[]) => void;

export interface NotieContextAttributes extends NotieSettings {
    success(message: string, ttl?: number): void;
    error(message: string, ttl?: number): void;
    warn(message: string, ttl?: number): void;
    info(message: string, ttl?: number): void;
    confirm(
        message: string,
        props?: Pick<NotieSettings, 'okBtnText' | 'cancelBtnText'>
    ): Promise<unknown>;
    force(
        message: string,
        props?: Pick<NotieSettings, 'okBtnText' | 'cancelBtnText'>
    ): Promise<unknown>;
}

export const NotieContext = React.createContext<NotieContextAttributes>({
    message: '',
    level: NotieLevel.INFO,
    okBtnText: '',
    cancelBtnText: '',
    success: () => null,
    error: () => null,
    warn: () => null,
    info: () => null,
    confirm: async () => Promise.resolve(''),
    force: async () => Promise.resolve('')
});

const DEFAULT_STATE: NotieSettings = {
    message: '',
    level: NotieLevel.INFO,
    okBtnText: 'OK',
    cancelBtnText: 'Cancel'
};

let alreadyLoaded = false;

export const DEFAULT_TIMEOUT = 5000;

export class Notie extends React.Component<any, NotieSettings> {
    private dialog: HTMLDialogElement | null = null;
    private timeout: number | null = null;
    private confirmResolve!: NotieCallback;
    private confirmReject!: NotieCallback;

    constructor(props: any) {
        super(props);

        this.state = {
            ...DEFAULT_STATE
        };

        if (alreadyLoaded) {
            throw new Error(
                '<Notie /> is already loaded somwhere in your code. Please include it only once.'
            );
        }

        alreadyLoaded = true;
    }

    componentWillUnmount() {
        if (this.timeout) {
            clearTimeout(this.timeout);
            this.timeout = null;
        }

        if (alreadyLoaded) alreadyLoaded = false;
    }

    showNotie = (settings: NotieSettings) => {
        if (this.forced) return;

        if (this.timeout) {
            clearTimeout(this.timeout);
            this.timeout = null;
        }

        if (this.dialog && this.dialog.open) {
            this.hideNotie(() => {
                this.showNotie(settings);
            });

            return;
        }

        this.setState(
            (state) => ({
                message: settings.message,
                level: settings.level,
                okBtnText: settings.okBtnText || state.okBtnText,
                cancelBtnText: settings.cancelBtnText || state.cancelBtnText
            }),
            () => {
                if (!this.dialog) return;

                if (this.forced) {
                    this.dialog.showModal();
                } else {
                    this.dialog.show();
                    this.timeout = window.setTimeout(() => {
                        this.hideNotie();
                        this.timeout = null;
                    }, settings.ttl || DEFAULT_TIMEOUT);
                }

                this.dialog.animate(
                    [
                        { transform: 'translate3d(0, -106%, 0)' },
                        { transform: 'translate3d(0, 0, 0)' }
                    ] as any[],
                    {
                        duration: 300,
                        direction: 'normal'
                    }
                );
            }
        );
    };

    hideNotie = (callback?: NotieCallback) => {
        if (!this.dialog) return;

        const a = this.dialog.animate(
            [
                { transform: 'translate3d(0, 0, 0)' },
                { transform: 'translate3d(0, -106%, 0)' }
            ] as any[],
            {
                duration: 300,
                direction: 'normal'
            }
        );

        a.onfinish = () => {
            if (this.dialog) this.dialog.close();
            if (typeof callback === 'function') callback();
            this.setState(DEFAULT_STATE);
        };
    };

    handleYes = () => {
        this.hideNotie(this.confirmResolve);
    };

    handleNo = () => {
        this.hideNotie(this.confirmReject);
    };

    handleDismiss = () => {
        if (this.forced) return;

        if (this.timeout) {
            clearTimeout(this.timeout);
            this.timeout = null;
        }

        if (this.dialog && this.dialog.open) {
            this.hideNotie();
        }
    };

    rootRef = (c: HTMLDialogElement) => {
        this.dialog = c;
    };

    get forced() {
        const { level } = this.state;

        return level === NotieLevel.CONFIRM || level === NotieLevel.FORCE;
    }

    showSuccess: NotieContextAttributes['success'] = (message, ttl) => {
        this.showNotie({
            ...DEFAULT_STATE,
            level: NotieLevel.SUCCESS,
            message,
            ttl
        });
    };

    showError: NotieContextAttributes['error'] = (message, ttl) => {
        this.showNotie({
            ...DEFAULT_STATE,
            level: NotieLevel.ERROR,
            message,
            ttl
        });
    };

    showWarn: NotieContextAttributes['warn'] = (message, ttl) => {
        this.showNotie({
            ...DEFAULT_STATE,
            level: NotieLevel.WARN,
            message,
            ttl
        });
    };

    showInfo: NotieContextAttributes['info'] = (message, ttl) => {
        this.showNotie({
            ...DEFAULT_STATE,
            level: NotieLevel.INFO,
            message,
            ttl
        });
    };

    showConfirm: NotieContextAttributes['confirm'] = async (message, props) => {
        // tslint:disable-next-line:no-inferred-empty-object-type
        return new Promise(
            (resolve, reject): void => {
                this.confirmResolve = resolve;
                this.confirmReject = reject;
                this.showNotie({
                    ...DEFAULT_STATE,
                    level: NotieLevel.CONFIRM,
                    ...props,
                    message
                });
            }
        );
    };

    showForce: NotieContextAttributes['force'] = async (message, props) => {
        // tslint:disable-next-line:no-inferred-empty-object-type
        return new Promise(
            (resolve, reject): void => {
                this.confirmResolve = resolve;
                this.confirmReject = reject;
                this.showNotie({
                    ...DEFAULT_STATE,
                    level: NotieLevel.FORCE,
                    ...props,
                    message
                });
            }
        );
    };

    render() {
        const { message, level } = this.state;

        return (
            <NotieContext.Provider
                value={{
                    ...this.state,
                    confirm: this.showConfirm,
                    error: this.showError,
                    info: this.showInfo,
                    success: this.showSuccess,
                    warn: this.showWarn,
                    force: this.showForce
                }}
            >
                {this.props.children}
                <dialog
                    className={notieContainer}
                    ref={this.rootRef}
                    style={{
                        backgroundColor: level in colors ? colors[level] : '#444'
                    }}
                    onClick={this.handleDismiss}
                >
                    <div className={notieMessage}>{message}</div>
                    {this.forced && (
                        <div className={notieChoices}>
                            <button
                                className={notieButton}
                                style={{
                                    backgroundColor:
                                        colors[
                                            level === NotieLevel.FORCE
                                                ? NotieLevel.ERROR
                                                : NotieLevel.SUCCESS
                                        ]
                                }}
                                onClick={this.handleYes}
                            >
                                {this.state.okBtnText}
                            </button>
                            {level === NotieLevel.CONFIRM && (
                                <button
                                    className={notieButton}
                                    style={{
                                        backgroundColor: colors[NotieLevel.ERROR]
                                    }}
                                    onClick={this.handleNo}
                                >
                                    {this.state.cancelBtnText}
                                </button>
                            )}
                        </div>
                    )}
                </dialog>
            </NotieContext.Provider>
        );
    }
}
