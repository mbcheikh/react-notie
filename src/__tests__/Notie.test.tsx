import * as React from 'react';
import { mount, ReactWrapper } from 'enzyme';

import { Notie, DEFAULT_TIMEOUT } from '../Notie';

jest.useFakeTimers();

describe('<Notie />', () => {
    let provider: ReactWrapper<{}, {}, Notie> | null;

    beforeEach(() => {
        provider = mount(<Notie />);
    });

    afterEach(() => {
        if (provider) {
            provider.unmount();
            provider = null;
        }
    });

    test('initial render', () => {
        expect(provider).toMatchSnapshot();
    });

    test('success message', () => {
        if (!provider) return;

        provider.instance().showSuccess('Success!');
        provider.update();
        expect(provider).toMatchSnapshot();
        // expect(provider.find(notieContainer).prop('level')).toBe(NotieLevel.SUCCESS);
        jest.runTimersToTime(DEFAULT_TIMEOUT + 100);
        provider.update();
        expect(provider).toMatchSnapshot();
    });

    test('warn message', () => {
        if (!provider) return;

        provider.instance().showWarn('Warn!');
        provider.update();
        expect(provider).toMatchSnapshot();
        // expect(provider.find(notieContainer).prop('level')).toBe(NotieLevel.WARN);
        jest.runTimersToTime(DEFAULT_TIMEOUT + 100);
        provider.update();
        expect(provider).toMatchSnapshot();
    });

    test('Error message', () => {
        if (!provider) return;

        provider.instance().showError('Error!');
        provider.update();
        expect(provider).toMatchSnapshot();
        // expect(provider.find('notieContainer').prop('level')).toBe(NotieLevel.ERROR);
        jest.runTimersToTime(DEFAULT_TIMEOUT + 100);
        provider.update();
        expect(provider).toMatchSnapshot();
    });

    test('info message', () => {
        if (!provider) return;

        provider.instance().showInfo('Info!');
        provider.update();
        expect(provider).toMatchSnapshot();
        // expect(provider.find(notieContainer).prop('level')).toBe(NotieLevel.INFO);
        jest.runTimersToTime(DEFAULT_TIMEOUT + 100);
        provider.update();
        expect(provider).toMatchSnapshot();
    });

    // test.skip('consecutive calls', () => {
    //     const instance = provider.instance();
    //     instance.info('Info!');
    //     provider.update();
    //     expect(provider.find(components.NotieContainer).prop('level')).toBe('SUCCESS');
    //     expect(provider.find(components.NotieContainer).prop('active')).toBe(true);
    //     expect(provider).toMatchSnapshot();
    //     jest.runTimersToTime(2000);
    //     provider.update();
    //     instance.error('Error!');

    //     /*
    //         Must be triggered manually since `transitionend` event
    //         is not being trigged when using `dispatchEvent` on root
    //         element.
    //     */
    //     instance.transitionendCallback();
    //     expect(instance.transitionendCallback).toBe(null);
    //     provider.update();

    //     expect(provider.find('.react-notie-level--error').length).toBe(1);
    //     expect(provider.find('.react-notie--active').length).toBe(1);
    //     expect(provider).toMatchSnapshot();
    // });

    // test.skip('cleanup on unmount', () => {
    //     const instance = provider.instance();
    //     actions.notieInfo('Info!');
    //     actions.notieInfo('Info!');
    //     jest.runTimersToTime(2000);
    //     provider.update();
    //     provider.unmount();
    //     expect(instance.transitionendCallback).toBe(null);
    // });

    // test('dismiss click', () => {
    //     actions.notieInfo('Info!');
    //     provider.update();
    //     expect(provider.find(components.NotieContainer).prop('level')).toBe('INFO');
    //     expect(provider.find(components.NotieContainer).prop('active')).toBe(true);
    //     expect(provider).toMatchSnapshot();
    //     provider.find(components.NotieContainer).simulate('click');
    //     provider.update();
    //     expect(provider.find(components.NotieContainer).prop('active')).toBe(false);
    //     expect(provider).toMatchSnapshot();
    // });

    // describe('confirm', () => {
    //     test('initial render', () => {
    //         actions.notieConfirm('Are you sure?');
    //         provider.update();
    //         expect(provider.find(components.NotieContainer).prop('level')).toBe('CONFIRM');
    //         expect(provider.find(components.NotieContainer).prop('active')).toBe(true);
    //         expect(provider).toMatchSnapshot();
    //     });

    //     test.skip('click yes', (done) => {
    //         const instance = provider.instance();
    //         const promise = instance.confirm('Are you sure?');

    //         provider.update();
    //         provider.find('.react-notie-choice--yes').simulate('click');
    //         instance.transitionendCallback();
    //         expect(instance.transitionendCallback).toBe(null);
    //         provider.update();

    //         promise.then(() => {
    //             expect(provider).toMatchSnapshot();
    //             expect(provider.find(components.NotieContainer).prop('active')).toBe(false);
    //             done();
    //         });
    //     });

    //     test.skip('click no', (done) => {
    //         const instance = provider.instance();
    //         const promise = instance.confirm('Are you sure?');

    //         provider.update();
    //         provider.find('.react-notie-choice--no').simulate('click');
    //         instance.transitionendCallback();
    //         expect(instance.transitionendCallback).toBe(null);
    //         provider.update();
    //         promise.catch(() => {
    //             expect(provider).toMatchSnapshot();
    //             expect(provider.find(components.NotieContainer).prop('active')).toBe(false);
    //             done();
    //         });
    //     });

    //     test.skip('no consecutive calls', () => {
    //         const instance = provider.instance();
    //         instance.confirm('Are you sure?');
    //         provider.update();
    //         expect(provider.find('.react-notie-level--confirm').length).toBe(1);
    //         expect(provider).toMatchSnapshot();
    //         instance.info('Info!');
    //         provider.update();
    //         expect(provider.find('.react-notie-level--confirm').length).toBe(1);
    //         expect(provider).toMatchSnapshot();
    //     });

    //     test.skip('Custom button text', () => {
    //         const instance = provider.instance();
    //         instance.confirm('Are you sure?', {
    //             yesBtnText: 'Hell Yeah!',
    //             noBtnText: 'Fuck NO!'
    //         });
    //         expect(provider).toMatchSnapshot();
    //     });
    // });
});
