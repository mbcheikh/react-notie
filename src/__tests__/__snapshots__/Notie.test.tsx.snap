// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notie /> Error message 1`] = `
.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
}

.emotion-1[open] {
  pointer-events: auto;
}

.emotion-0 {
  padding: 20px;
}

<Notie>
  <dialog
    className="emotion-1"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#ff4949",
      }
    }
  >
    <div
      className="emotion-0"
    >
      Error!
    </div>
  </dialog>
</Notie>
`;

exports[`<Notie /> Error message 2`] = `
.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
}

.emotion-1[open] {
  pointer-events: auto;
}

.emotion-0 {
  padding: 20px;
}

<Notie>
  <dialog
    className="emotion-1"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#ff4949",
      }
    }
  >
    <div
      className="emotion-0"
    >
      Error!
    </div>
  </dialog>
</Notie>
`;

exports[`<Notie /> info message 1`] = `
.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
}

.emotion-1[open] {
  pointer-events: auto;
}

.emotion-0 {
  padding: 20px;
}

<Notie>
  <dialog
    className="emotion-1"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#20a0ff",
      }
    }
  >
    <div
      className="emotion-0"
    >
      Info!
    </div>
  </dialog>
</Notie>
`;

exports[`<Notie /> info message 2`] = `
.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
}

.emotion-1[open] {
  pointer-events: auto;
}

.emotion-0 {
  padding: 20px;
}

<Notie>
  <dialog
    className="emotion-1"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#20a0ff",
      }
    }
  >
    <div
      className="emotion-0"
    >
      Info!
    </div>
  </dialog>
</Notie>
`;

exports[`<Notie /> initial render 1`] = `
.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
}

.emotion-1[open] {
  pointer-events: auto;
}

.emotion-0 {
  padding: 20px;
}

<Notie>
  <dialog
    className="emotion-1"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#20a0ff",
      }
    }
  >
    <div
      className="emotion-0"
    />
  </dialog>
</Notie>
`;

exports[`<Notie /> success message 1`] = `
.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
}

.emotion-1[open] {
  pointer-events: auto;
}

.emotion-0 {
  padding: 20px;
}

<Notie>
  <dialog
    className="emotion-1"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#13ce66",
      }
    }
  >
    <div
      className="emotion-0"
    >
      Success!
    </div>
  </dialog>
</Notie>
`;

exports[`<Notie /> success message 2`] = `
.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
}

.emotion-1[open] {
  pointer-events: auto;
}

.emotion-0 {
  padding: 20px;
}

<Notie>
  <dialog
    className="emotion-1"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#13ce66",
      }
    }
  >
    <div
      className="emotion-0"
    >
      Success!
    </div>
  </dialog>
</Notie>
`;

exports[`<Notie /> warn message 1`] = `
.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
}

.emotion-1[open] {
  pointer-events: auto;
}

.emotion-0 {
  padding: 20px;
}

<Notie>
  <dialog
    className="emotion-1"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#f5be4b",
      }
    }
  >
    <div
      className="emotion-0"
    >
      Warn!
    </div>
  </dialog>
</Notie>
`;

exports[`<Notie /> warn message 2`] = `
.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
}

.emotion-1[open] {
  pointer-events: auto;
}

.emotion-0 {
  padding: 20px;
}

<Notie>
  <dialog
    className="emotion-1"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#f5be4b",
      }
    }
  >
    <div
      className="emotion-0"
    >
      Warn!
    </div>
  </dialog>
</Notie>
`;
